File: 20.有效的括号.go
/*
 * @lc app=leetcode.cn id=20 lang=golang
 *
 * [20] 有效的括号
 */
package leetcode

// @lc code=start
func isValid(s string) bool {
	var stack []rune

	pair := map[rune]rune{
		'(': ')',
		'[': ']',
		'{': '}',
	}
	for _, v := range s {
		if v == '(' || v == '{' || v == '[' {
			stack = append(stack, v)
		} else {
			if len(stack) == 0 {
				return false
			}
			if v == pair[stack[len(stack)-1]] {
				stack = stack[0 : len(stack)-1]
			} else {
				return false
			}
		}
	}

	return len(stack) == 0
}

// @lc code=end

File: rust/20.有效的括号.rs
/*
 * @lc app=leetcode.cn id=20 lang=rust
 *
 * [20] 有效的括号
 */

// @lc code=start
impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack: Vec<char> = vec![];

        let match_closures = |x| -> char {
            match x {
                ']' => '[',
                '}' => '{',
                ')' => '(',
                _ => '1',
            }
        };

        for c in s.chars() {
            if stack.is_empty() {
                stack.push(c);
            } else {
                let a = *stack.last().unwrap();
                if a.eq(&match_closures(c)) {
                    stack.pop();
                } else {
                    stack.push(c);
                }
            }
        }

        if stack.is_empty() {
            true
        } else {
            false
        }
    }
}
// @lc code=end
